firmware/port.c

```diff
@@ -211,6 +302,61 @@ void port_exec_async_complete(PortData* p, ExecStatus s) {
     port_step(p);
 }
 
+void usart_test_send(SercomId id, uint8_t* buf, uint32_t buf_len){
+    if (!(sercom(id)->USART.INTFLAG.reg & SERCOM_USART_INTFLAG_DRE)) {
+        invalid();
+        return;
+    }
+
+    while (sercom(id)->USART.SYNCBUSY.reg) {
+        // wait until not busy
+    }
+
+    for (uint32_t i = 0; i < buf_len; i++){
+        // write one byte
+        sercom(id)->USART.DATA.reg = buf[i];
+
+        while (!(sercom(id)->USART.INTFLAG.reg & SERCOM_USART_INTFLAG_TXC)) {
+            // wait until byte is written
+        }
+    }
+}
+
+void uart_send_data(PortData *p){
+    if (uart_buf.buf_len > 0) {
+        // copy data into port and send it
+        // buffer is full, send off data
+        // p->state = EXEC_DONE;
+
+        // pad 2 bytes at the beginning
+        // 1st byte indicates uart rx
+        // 2nd byte indicates uart rx number. 
+        // this also means rx number has to be <255
+        // p->reply_len += uart_buf.buf_len;
+        if (p->uart_data_pos == 255) {
+            // if this is the first time that uart data is 
+            // getting put into reply buf, pad by 2.
+            p->uart_data_pos = p->reply_len; // keep track of where we write uart data headers
```

Comment:
I had `uart_data_pos` in here because we could fire this off to port_send (and have data already copied to the reply buffer), but not actually be done with sending off that data before `uart_send_data` is called again. So basically:

1. some uart data comes in, uart data buffer gets filled up or the timer hits and we call `uart_send_data`
2. data gets copied into reply buf. `port_send` is called to pass this across the bridge
3. before the bridge transfer can complete, more uart data comes in and `uart_send_data` is called again.
4. we append the new uart data in to the current reply_buf and increment the length of data that the uart packet contains.
5. once the bridge dma completes, uart_data_pos is reset.